Start at running ILLink to find dead code in Corelib

danmosemsft opened the pull request:
A start of looking at what ILLink discards from Corelib. I compared the results using ApiReviewer. Apparently there aren't many dead private members, and many of those actually need rooting in the xml configuration file.

I didn't finish the privates but there weren't many left - presumably because BCLRewriter was doing a good job.

However there are very many apparently dead internals (~1000 = around 100-200KB of ~2800KB, allowing for stuff that's Unix only) that are being rooted by the `InternalsVisibleTo` attributes. Temporarily removing the attributes makes them show up. Most could be removed.

This at least documents how to continue the process, if we want to trim more size.

danmosemsft commented:
Here's what it believes is still dead, mostly internals. Not filtered out: stuff the VM or friend assemblies need; Unix or release stuff.

https://gist.github.com/danmosemsft/0cf6a2d9b0970f72fe33f67c83224998

jkotas commented:
> https://gist.github.com/danmosemsft/0cf6a2d9b0970f72fe33f67c83224998

I have looked over this list. The few methods in `class Win32Native` is obvious dead code - you can add it to this change. But majority looks like false positives. A lot direct and indirect dependencies of the interop implementation in the VM in the list.

jkotas commented on src/mscorlib/src/System/AppDomainSetup.cs:
```diff
@@ -62,17 +61,11 @@ internal enum LoaderInformation
         private Dictionary<string, object> _CompatFlags;
 
-        [OptionalField(VersionAdded = 5)] // This was added in .NET FX v4.5
         private String _TargetFrameworkName;
```
I think this can be deleted too. It is always null, never set to anything.

danmosemsft commented:
Fixed those and removed some other randomly selected dead code from scanning the list. Quite a bit of file related stuff is Unix only, left that.

Since as you say there is less here than it seemed I don't expect to spend much/any more time on going through it right now.

jkotas commented on src/mscorlib/shared/System/IO/PathInternal.Windows.cs:
```diff
@@ -437,11 +437,6 @@ internal static bool IsDirectoryOrVolumeSeparator(char ch)
             return IsDirectorySeparator(ch) || VolumeSeparatorChar == ch;
         }
 
-        internal static string TrimEndingDirectorySeparator(string path) =>
```
This is used in `shared\Microsoft\Win32\SafeHandles\SafeFileHandle.Unix.cs`

jkotas commented:
arm tests failing

```
Assert failure(PID 984 [0x000003d8], Thread: 984 [0x03d8]): Consistency check failed: Managed object size does not match unmanaged object size
man: 0x14, unman: 0x18, Name: System.AppDomainSetup
```

jkotas commented on src/mscorlib/src/Microsoft/Win32/Win32Native.cs:
```diff
@@ -596,16 +578,6 @@ internal static bool GlobalMemoryStatusEx(ref MEMORYSTATUSEX buffer)
         internal const int FIND_FROMSTART = 0x00400000; // look for value in source, starting at the beginning
         internal const int FIND_FROMEND = 0x00800000; // look for value in source, starting at the end
```
These Nls constants can be deleted too.

danmosemsft replied:
done. Note that ILLink will never tell us about these, as it shows all consts as "removed". It would be nice to make it more useful for dead code removal eg to not show default empty ctors, inlined constants,. .

jkotas commented on src/mscorlib/src/System/AppDomainSetup.cs:
```diff
@@ -196,11 +185,7 @@ public String TargetFrameworkName
-            set
-            {
-                _TargetFrameworkName = value;
+                return null;
```
This cleanup can go one level up to AppContext.

There is inconsistency between CoreCLR and CoreRT actually: In CoreCLR, AppContext.TargetFrameworkName will always return null. On CoreRT/ProjectN, it will return attribute value from main exe https://github.com/dotnet/corert/blob/ff7eba504516fa0bc0a50357ad559c19b45d1d51/src/System.Private.CoreLib/src/System/AppContext.cs#L45

danmosemsft replied:
Interesting. I will break out an issue since this will cause AppContext to parse ".NETCoreApp,Version=v2.0" and have an identifier and version instead of empty/null as it currently has.

Accordingly I will leave this return null here since the easiest fix would paste `                return Assembly.GetEntryAssembly()?.GetCustomAttribute<TargetFrameworkAttribute>()?.FrameworkName;` into AppDomainSetup here - that's where AppContextDefaultValues gets it form.

AlexGhiondea replied:
Some of those fields are set from the runtime -- I am not sure this should be removed.

danmosemsft commented:
@dotnet/dnceng can you figure out why the perf tests  correctness jobs haven't started running after 4 hours wait?

danmosemsft commented:
@dotnet-bot test Windows_NT x64 CoreCLR Perf Tests Correctness  please
@dotnet-bot test Windows_NT x86 CoreCLR Perf Tests Correctness  please

mmitche commented:
@danmosemsft Looking into it

mmitche commented:
https://github.com/dotnet/core-eng/issues/1462

This node (and only this one) is allocating but not connecting.  The script that Azure is supposed tor run isn't running.

mmitche commented:
@danmosemsft Updated the bug with some more info.  It has something to do with Azure changing custom script extension versions.  Either something went wrong on their side, or the VM images that aren't connecting have something wrong with them.  I'm trying to track it down.

danmosemsft commented:
Well it works now thanks @mmitche

danmosemsft commented:
@AlexGhiondea

danmosemsft commented:
@AlexGhiondea as mentioned
